doc/holochain_101/src/dpki.md

```diff
@@ -0,0 +1,54 @@
+
+In designing a solution for DPKI we recognize that this is a complex and difficult enough problem that any solution will need to evolve, and in fact there will be multiple solutions necessary for different contexts.  Thus, we have built into the Holochain conductor a simple interface for the fundamental needed functions, e.g. creating new keys when installing a DNA for the first time, that can then be implemented by specialized DPKI applications.  Furthermore we've implemented a reference implementation of a Holochain based DPKI application, which we call DeepKey.
+
+## DeepKey
+
+To deliver on the basics of Distributed Public Key Infrastructure, we need a way to generate keys of various types (revocation, identity, encryption, signing) from seeds, and we need to be able to generate such seeds from primary seeds, so that a human agent can create related "device agents" provably under their control.
+
+After studying a number of uses cases, including initial sign-up, key revocation, etc, the central insight we came to was the need to create a Hierarchical Deterministic Key generation system, based on a Primary Seed, from which additional seeds can be generated which then are in turn used to actually generate many key-pairs.  This allows us, by-convention, to use the first seed generated by the Primary seed as the seed for revocation keys, and subsequent seeds as seeds for keys of separate Holochain devices that can be proven to be under the control of the holder of Primary Seed.
+
+DeepKey
+
+TODO: merge the various docs we developed to explain DeepKey here.
+- https://medium.com/holochain/part-2-holochain-holo-accounts-cryptographic-key-management-and-deepkey-bf32ee91af65
+- https://hackmd.io/UbfvwQdJRKaAHI9Xa7F3VA?view
+- https://hackmd.io/8c8rZCyaTTqH_7TIBVtEUQ
+- https://hackmd.io/oobu0sKMSMadLXza4rHY_g
+
+## DPKI Implementation Technical Details
+
+### Keystore
+For each Holochain DNA instance, the Conductor maintains a Keystore, which holds "secrets" (seeds and keys) needed for cryptographic signing and encrypting. Each of the secrets in the Keystore is associated with a string which is a handle needed when using that secret for some cryptographic operation.  Our cryptographic implementation is based on libsodium, and the seeds use their notions of context and index for key derivation paths.  This implementation allows DNA developers to securely call cryptographic functions from WASM which will be executed in the conductor's secure memory space when actually doing the cryptographic processing.
+
+Decrypting a secret from the keystore invokes a passphrase manager service, which is used to collect the passphrase from some end-user.  This service is implementation specific.  Currently we have implementations for command-line passphrase collection for use in the `hc keyget` command, and also an implementation that routes through the same JSON RPC pathway for other admin commands sent to the conductor for other admin actions.
```

Comment:
We don't have an implementation for passphrase through JSON RPC yet...
Also, this should it should be `keygen` instead of `keyget`, right?
```suggestion
Decrypting a secret from the keystore invokes a passphrase manager service, which is used to collect the passphrase from some end-user.  This service is implementation specific.  Currently we have implementations for command-line passphrase collection for use in the `hc keygen` command, and also a command-line implementation within the conductor.
```